File: scala/custom-set/src/main/scala/CustomSet.scala

class CustomSet[+T](list: Seq[T]){
  val _v = list.distinct
}
object CustomSet {
  def fromList[T](list: Seq[T]) = new CustomSet[T](list)
  def empty[T](set: CustomSet[T]) = set._v.isEmpty
  def member[T](set: CustomSet[T], elem: T) = set._v.contains(elem)
  def isSubsetOf[T]( s1: CustomSet[T], s2: CustomSet[T] ) = s1._v.forall(s2._v.contains(_))
  def isDisjointFrom[T]( s1: CustomSet[T], s2: CustomSet[T] ) = !s1._v.exists(s2._v.contains(_))
  def isEqual[T]( s1: CustomSet[T], s2: CustomSet[T] ) = s1._v.length == s2._v.length && isSubsetOf(s1, s2)
  def insert[T]( set: CustomSet[T], elem: T ): CustomSet[T] = new CustomSet[T](set._v ++ Seq(elem))
  def intersection[T]( s1: CustomSet[T], s2: CustomSet[T] ) = new CustomSet[T](s1._v.filter(s2._v.contains(_)))
  def union[T]( s1: CustomSet[T], s2: CustomSet[T] ) = new CustomSet[T](s1._v ++ s2._v)
  def difference[T]( s1: CustomSet[T], s2: CustomSet[T] ) = new CustomSet[T](s1._v diff s2._v)


}

File: javascript/custom-set/custom-set.js
export class CustomSet {
    constructor(arr = []) {
        this.lst = arr.sort((a, b) => a - b);
    }

    empty = () => !this.lst.length;

    contains = element => this.lst.includes(element);

    add = element => {
        if (this.lst.indexOf(element) < 0) this.lst.push(element);
        this.lst = this.lst.sort((a, b) => a - b);
        return this;
    };

    subset = anotherSet => {
        if (!this.lst.length) return true;
        return this.lst.every(item => anotherSet.lst.indexOf(item) > -1);
    };

    disjoint = anotherSet => anotherSet.lst.every(item => this.lst.indexOf(item) === -1);

    eql = anotherSet => anotherSet.lst.toString() === this.lst.toString();

    union = anotherSet => {
        const joined = [...this.lst, ...anotherSet.lst];
        this.lst = joined.filter((item, index) => joined.indexOf(item) === index).sort((a, b) => a - b);
        return this;
    };

    intersection = anotherSet => {
        this.lst = this.lst.reduce((total, current) => {
            if (anotherSet.lst.indexOf(current) > -1) total.push(current);
            return total;
        }, []);
        return this;
    };

    difference = anotherSet => {
        this.lst = this.lst.reduce((total, current) => {
            if (anotherSet.lst.indexOf(current) < 0) total.push(current);
            return total;
        }, []);
        return this;
    };
}
